internal/serialization/default_compact_reader.go

```diff
@@ -0,0 +1,987 @@
+		panic(r.unexpectedFieldKind(fd.fieldKind, fieldName))
+	}
+	return *fd
+}
+
+func (r *DefaultCompactReader) getFixedSizePosition(fd FieldDescriptor) int32 {
+	primitiveOffset := fd.offset
+	return primitiveOffset + r.dataStartPosition
+}
+
+func (r *DefaultCompactReader) unknownField(fieldName string) error {
+	return ihzerrors.NewSerializationError(fmt.Sprintf("Unknown field name '%s' for %s", fieldName, r.schema.ToString()), nil)
+}
+
+func (r *DefaultCompactReader) unexpectedFieldKind(actualFieldKind pserialization.FieldKind, fieldName string) error {
+	return ihzerrors.NewSerializationError(fmt.Sprintf("Unexpected field kind '%d' for field %s", actualFieldKind, fieldName), nil)
+}
+
+func (r *DefaultCompactReader) unexpectedNullValue(fieldName, methodSuffix string) error {
+	return ihzerrors.NewSerializationError(fmt.Sprintf("Error while reading %s. nil value cannot be read via read%s methods. Use readNullable%s instead", fieldName, methodSuffix, methodSuffix), nil)
+}
+
+func (r *DefaultCompactReader) unexpectedNullValueInArray(fieldName, methodSuffix string) error {
+	return ihzerrors.NewSerializationError(fmt.Sprintf("Error while reading %s. nil value cannot be read via readArrayOf%s methods. Use readArrayOfNullable%s instead", fieldName, methodSuffix, methodSuffix), nil)
+}
+
+func (r *DefaultCompactReader) readVariableSize(fd FieldDescriptor, reader Reader) interface{} {
+	currentPos := r.in.position
+	defer r.in.SetPosition(currentPos)
+	position := r.readVariableSizeFieldPosition(fd)
+	if position == NULL_OFFSET {
+		return nil
+	}
+	r.in.SetPosition(position)
+	return reader(r.in)
+}
+
+func (r *DefaultCompactReader) readVariableSizeSlice(fd FieldDescriptor, reader SliceReader) []interface{} {
+	currentPos := r.in.position
+	defer r.in.SetPosition(currentPos)
+	position := r.readVariableSizeFieldPosition(fd)
+	if position == NULL_OFFSET {
+		return nil
+	}
+	r.in.SetPosition(position)
+	return reader(r.in)
+}
+
+func (r *DefaultCompactReader) readVariableSizeAsNonNull(fd FieldDescriptor, reader Reader, methodSuffix string) interface{} {
+	value := r.readVariableSize(fd, reader)
+	if value == nil {
+		panic(r.unexpectedNullValue(fd.fieldName, methodSuffix))
+	}
+	return value
+}
+
+func (r *DefaultCompactReader) readVariableSizeFieldPosition(fd FieldDescriptor) int32 {
+	index := fd.index
+	offset := r.offsetReader.getOffset(r.in, r.variableOffsetsPosition, index)
+	if offset == NULL_OFFSET {
+		return NULL_OFFSET
+	} else {
+		return offset + r.dataStartPosition
+	}
+}
+
+func (r *DefaultCompactReader) readBooleanBits(inp *ObjectDataInput) []bool {
+	len := inp.ReadInt32()
+	if len == NULL_ARRAY_LENGTH {
+		return nil
+	}
+	if len == 0 {
+		return make([]bool, 0)
+	}
+	values := make([]bool, len)
+	index := 0
+	currentByte := inp.ReadByte()
+	for i := int32(0); i < len; i++ {
+		if index == BitsInAByte {
+			index = 0
+			currentByte = inp.ReadByte()
+		}
+		result := ((currentByte >> index) & 1) != 0
```

Comment:
But we need a loop here anyway that is running through from 0 to `index`.  Note that the loop is not from 0 to 7. If it was, we would be able to what you said.

Reply:
Maybe we can loop for length/8 and do the operation described above with `i+0,i+1,i+2...`. I think it's doable but it's okay if you don't like it 😄

Reply:
Yeah we need another loop for the reminder of `len % 8`. It will be less readable for sure. What is the benefit of your approach btw? I think performance should not change much because length is at most a 32 bit integer. But I am not sure.

Reply:
I thought it would be more performant but you are right, with the reminder it would be hard to read. Thanks for the answer